src/router/builder/single.rs

```diff
@@ -0,0 +1,313 @@
+use router::route::dispatch::{PipelineHandleChain, DispatcherImpl};
+use handler::{Handler, NewHandler};
+
+/// Describes the API for defining a single route, after determining which request paths will be
+/// dispatched here. The API here uses chained function calls to build and add the route into the
+/// `RouterBuilder` which created it.
+///
+/// # Examples
+///
+/// ```rust
+/// # extern crate gotham;
+/// # extern crate hyper;
+/// # use hyper::Response;
+/// # use gotham::state::State;
+/// # use gotham::router::Router;
+/// # use gotham::router::builder::*;
+/// # use gotham::middleware::pipeline::new_pipeline;
+/// # use gotham::middleware::session::NewSessionMiddleware;
+/// # use gotham::router::route::dispatch::{new_pipeline_set, finalize_pipeline_set};
+/// fn my_handler(_: State) -> (State, Response) {
+///     // Handler implementation elided.
+/// #   unimplemented!()
+/// }
+/// #
+/// # fn router() -> Router {
+/// #   let pipelines = new_pipeline_set();
+/// #   let (pipelines, default) =
+/// #       pipelines.add(new_pipeline().add(NewSessionMiddleware::default()).build());
+/// #
+/// #   let pipelines = finalize_pipeline_set(pipelines);
+/// #
+/// #   let default_pipeline_chain = (default, ());
+///
+/// build_router(default_pipeline_chain, pipelines, |route| {
+///     route.get("/request/path") // <- This value implements `DefineSingleRoute`
+///          .to(my_handler);
+/// })
+/// # }
+/// # fn main() { router(); }
+/// ```
+pub trait DefineSingleRoute {
+    /// Directs the route to the given `Handler`, automatically creating a `NewHandler` which
+    /// copies the `Handler`. This is the easiest option for code which is using bare functions as
+    /// `Handler` functions.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # extern crate gotham;
+    /// # extern crate hyper;
+    /// # use hyper::Response;
+    /// # use gotham::state::State;
+    /// # use gotham::router::Router;
+    /// # use gotham::router::builder::*;
+    /// # use gotham::middleware::pipeline::new_pipeline;
+    /// # use gotham::middleware::session::NewSessionMiddleware;
+    /// # use gotham::router::route::dispatch::{new_pipeline_set, finalize_pipeline_set};
+    /// fn my_handler(_: State) -> (State, Response) {
+    ///     // Handler implementation elided.
+    /// #   unimplemented!()
+    /// }
+    /// #
+    /// # fn router() -> Router {
+    /// #   let pipelines = new_pipeline_set();
+    /// #   let (pipelines, default) =
+    /// #       pipelines.add(new_pipeline().add(NewSessionMiddleware::default()).build());
+    /// #
+    /// #   let pipelines = finalize_pipeline_set(pipelines);
+    /// #
+    /// #   let default_pipeline_chain = (default, ());
+    ///
+    /// build_router(default_pipeline_chain, pipelines, |route| {
+    ///     route.get("/request/path").to(my_handler);
+    /// })
+    /// # }
+    /// # fn main() { router(); }
+    /// ```
+    fn to<H>(self, handler: H)
```

Comment:
This is purely discussion.

I am wondering if the naming here should be `to/to_handler` as opposed to `to/to_new_handler` as I could easily see folks skipping over the concept of the `NewHandler` type (at least initially) and this getting confusing.

Brevity in the builder is also a nice side effect of this if we decided to change.

I understand this makes the builder not quite "pure" in naming, but worth considering.

Reply:
I disagree, for the same reason you mentioned. Though, instead of calling it "pure", I'd just call it "accurate".

`Handler + Copy` is the "default" here, which completely obviates the need for `NewHandler` in the trait bounds for `to<H>`. Having `to` take a `Handler + Copy`, and `to_handler` take a `NewHandler` seems like it would be _far_ more confusing for people who've learned what a `NewHandler` is.

For the 99.9% case, and especially new Gotham users, I see `to_new_handler` not being used at all. It's just there for completeness, in being able to control the exact `NewHandler` which is used for the route.

Reply:
I can live with this reasoning, consider this resolved.